lib/Support/LoweringOptions.cpp

```diff
@@ -0,0 +1,113 @@
+  parse(options, errorHandler);
+}
+
+LoweringOptions::LoweringOptions(mlir::ModuleOp module) : LoweringOptions() {
+  getFromAttribute(module);
+}
+
+void LoweringOptions::parse(StringRef text, ErrorHandlerT errorHandler) {
+  while (!text.empty()) {
+    // Remove the first option from the text.
+    auto split = text.split(",");
+    auto option = split.first.trim();
+    text = split.second;
+    if (option == "") {
+      // Empty options are fine.
+    } else if (option == "noAlwaysFF") {
+      useAlwaysFF = false;
+    } else {
+      errorHandler(llvm::Twine("unknown style option \'") + option + "\'");
+      // We continue parsing options after a failure.
+    }
+  }
+}
+
+std::string LoweringOptions::toString() const {
+  std::string options = "";
+  if (!useAlwaysFF)
+    options += "noAlwaysFF";
+  return options;
+}
+
+void LoweringOptions::setAsAttribute(ModuleOp module) {
+  module->setAttr("circt.loweringOptions",
+                  StringAttr::get(module.getContext(), toString()));
+}
+
+void LoweringOptions::getFromAttribute(ModuleOp module) {
+  if (auto styleAttr =
+          module->getAttrOfType<StringAttr>("circt.loweringOptions")) {
+    parse(styleAttr.getValue(), [&](Twine error) { module.emitError(error); });
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Command Line Option Processing
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Commandline parser for LoweringOptions.  Delegates to the parser
+/// defined by LoweringOptions.
+struct LoweringOptionsParser : public llvm::cl::parser<LoweringOptions> {
+
+  LoweringOptionsParser(llvm::cl::Option &option)
+      : llvm::cl::parser<LoweringOptions>(option) {}
+
+  bool parse(llvm::cl::Option &option, StringRef argName, StringRef argValue,
+             LoweringOptions &value) {
+    bool failed = false;
+    value.parse(argValue, [&](Twine error) { failed = option.error(error); });
+    return failed;
+  }
+};
+
+/// Commandline arguments for verilog emission.  Used to dynamically register
+/// the command line arguments in multiple tools.
+struct LoweringCLOptions {
+  llvm::cl::opt<LoweringOptions, false, LoweringOptionsParser> loweringOptions{
+      "circt-lowering-options", llvm::cl::desc("Style options"),
+      llvm::cl::value_desc("option")};
+};
+} // namespace
+
+/// The staticly initialized command line options.
+static llvm::ManagedStatic<LoweringCLOptions> clOptions;
+
+void circt::registerLoweringCLOptions() { *clOptions; }
+
+void circt::applyLoweringCLOptions(ModuleOp module) {
+  // If the command line options were not registered in the first place, there
+  // is nothing to parse.
+  if (!clOptions.isConstructed())
+    return;
+
+  // If an output style is applied on the command line, all previous options are
+  // discarded.
+  if (clOptions->loweringOptions.getNumOccurrences()) {
+    clOptions->loweringOptions.setAsAttribute(module);
```

Comment:
I'm not sure that this has the right effect.  If I specify "attr1=42" on the command line, but the module has "attr2=17" then I'll get "attr1=42,attr2=defaultvalue" instead of "attr1=42,attr2=17".

I'd suggest simplifying this logic a bit: just have a cl::opt<string>, then in applyLoweringCLOptions you can construct a LoweringOptions from the module, then parse in the string (as specified on the command line) then set the aggregate back into the module.  I think that will give you the right "default to what is in the module, but allow deviations from that on the command line" behavior.  It would also eliminate the need for a custom option parser

Reply:
The issue with that is you can't turn off `noAlwaysFF` from the command line. E. g. the module has `noAlwaysFF`, and the user tries to turn off this option on the command line with the string "", which does not override anything.  What we want is to allow the user to specify `noAlwaysFF=false`.  Since we share the parser between the command line and the module attributes, they both would accept this format.

This is the 3-state boolean thing I mentioned before:
 `unset` -> leave the value alone, 
 `true` -> set the option to true,
 `false` -> set the option to false. 
This would allow us to override the individual field on the command line. 
 
 This string could be "canonicalized" on the module by reading it in to an options struct and then recreating the string, which I think works pretty well.
 
 Let me know if you think this is a good idea and I can get started on it.

Reply:
I see.  Ok, please land this patch as is and we can continue to iterate on it.  I think your tristate thing is a good idea for boolean options.  We can support `setting=true` and `setting=false` and use `setting` as an alias for `setting=true` in boolean case.

Other settings will have specific values `linelength=42` so this will only come up with boolean options.